### PR TITLE
Speed up `Tensor::has_names` for unnamed tensors

### PR DESCRIPTION
**Pull request by zou3519**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#31436 Speed up `Tensor::has_names` for unnamed tensors**

Tensor::has_names is slower than it should be for unnamed tensors
because of the following:
- it always tries to access the TLS for NamesMode. Unnamed tensors don't
need to peek at NamesMode to determine if they have names or not.
- There is some virtual function being called because TensorImpl is in
c10 and NamedTensorMeta is in libtorch.

This PR short-circuits Tensor::has_names for unnamed tensors by
checking if the underlying TensorImpl hold a pointer to NamedTensorMeta
or not. If the NamedTensorMeta is nullptr; then the tensor is definitely
unnamed.

Benchmarks:
- I have a dedicated benchmarking machine where I isolate a single CPU
and make sure it runs at a fixed frequency.
- I benchmarked torch.add, which calls `tensor::has_names` three times.
- The TL;DR is that torch.add between size-1 unnamed tensors gets sped up
~200ns after this change which is a 9% improvement.
- Before, on my machine:
https://gist.github.com/zou3519/dfd648a1941d584711d850754e0694bc
- After on my machine:
https://gist.github.com/zou3519/e78f0d8980b43d0d9c3e3e78ecd0d4d5

Test Plan:
- run tests

Differential Revision: [D19166510](https://our.internmc.facebook.com/intern/diff/D19166510)

### CODE REVIEW

**Comment by kostmo** [comment]
## :pill: CircleCI build failures summary and remediations
As of commit 86e3bbc:
* **1/1** broken upstream at merge base 47766e6 (see [grid view](https://dr.pytorch.org/master-timeline.html?count=1&sha1=47766e648ff16d4b1175d04710caed566de14ab4&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
  <details>
  <summary>
  You may want to rebase on the <code>viable/strict</code> branch (expand for instructions)
  </summary>
  
  If your commit is newer than `viable/strict`, you can try basing on an older, stable commit:
  ```
  git fetch viable/strict
  git rebase --onto viable/strict $(git merge-base origin/master HEAD)
  ```
  If your commit is older than `viable/strict`:
  ```
  git fetch viable/strict
  git rebase viable/strict
  ```
  Check out the [recency history](https://dr.pytorch.org/master-viable-commits.html) of this "viable master" tracking branch.
  
  </details>
  
* **0/1** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=86e3bbc91d2cc310b0c8a80c3177e4ad6d05bdf6).

### :construction: 1 upstream failure recognized by patterns:
These builds matched patterns, but were probably caused by upstream breakages:
* [pytorch_xla_linux_xenial_py3_6_clang7_build](https://dr.pytorch.org/build-details.html?build_id=76831384)



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

**Review by albanD — APPROVED** [review]
